### ISSUE TITLE
`clippy::zero_prefixed_literal` emitted incorrectly when a non-zero literal is emitted from a macro with a span of a `0` literal

### ISSUE THREAD

**Issue by dhedey** [rust-lang/rust-clippy#14199]
## Context

The [zero_prefixed_literal](https://rust-lang.github.io/rust-clippy/master/index.html#zero_prefixed_literal) lint triggered with the token string `#(x = 0)` inside a macro.

Specifically, the macro is a compile-time interpreter (PR here: https://github.com/dhedey/preinterpret/pull/1).

## Details

```rust
#[test]
fn test_while() {
    assert_preinterpret_eq!({
        #(x = 0)
        [!while! #x < 5 { #(x += 1) }]
        #x
    }, 5);
}
```

Error message:
```text
warning: this is a decimal constant
  --> tests/control_flow.rs:53:15
   |
53 |         #(x = 0)
   |               ^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#zero_prefixed_literal
   = note: `#[warn(clippy::zero_prefixed_literal)]` on by default
help: if you mean to use a decimal constant, remove the `0` to avoid confusion
   |
53 -         #(x = 0)
53 +         #(x = )
   |
help: if you mean to use an octal constant, use `0o`
   |
53 |         #(x = 0o)
   |               ~~
```

## Hypothesis

I _think_ that the issue is that there is a literal 5 output from the macro with a source span from a 0 literal.

Indeed:
* Changing ` #(x = 0)` to `#(x = 1)` so that the source literal doesn't start with `0` avoids the lint
* Changing `#x < 5` to `#x < 0` so that it outputs `0` avoids the lint

### Lint Name

[zero_prefixed_literal](https://rust-lang.github.io/rust-clippy/master/index.html#zero_prefixed_literal)

### Reproducer

(See above)

### Version

```text
rustc 1.84.0 (9fc6b4312 2025-01-07)
binary: rustc
commit-hash: 9fc6b43126469e3858e2fe86cafb4f0fd5068869
commit-date: 2025-01-07
host: aarch64-apple-darwin
release: 1.84.0
LLVM version: 19.1.5
```

**Comment by samueltardieu** [comment]
This could happen if a procedural macros emitted a number while improperly reusing the span of another. Without a self-contained reproducer, it is hard to tell.

**Comment by dhedey** [comment]
> This could happen if a procedural macros emitted a number while improperly reusing the span of another. Without a self-contained reproducer, it is hard to tell.

That's correct:
* When this ticket was raised, the macro was outputting the literal `5` with a span attempting to join the `0` and each `1` which was added. But `span::join()` isn't stable, so on stable rust, it just returns the first span `0` - which I assume triggers this lint.
* Since then, the macro has been altered so that the output from a variable takes the output span from the variable tokens (in this case, `#x`), so this bug no longer flares up.

So you're right that this issue is caused by a number reusing the span of another, and I'm happy to close this ticket!

That said, there is something slightly weird about clippy's behaviour here: it's identifying the value of the literal is non-zero (at use time?), but reading the content of the literal from its span and not noticing they are different values.

I imagine that this is very unlikely occurrence and it's not worth any time to invest in this for now, but if other people hit this in future, perhaps clippy could:
* Detect the value and literal disagree , and output a new lint for that e.g. `clippy::misapplied_literal_span`
* Only if they agree, then output `clippy::zero_preflixed_literal`
